### PR TITLE
Deprecate DSA harder (cont. from #13187)

### PR DESCRIPTION
**Pull request by levitte**
This deprecates all functions that deal with the type DSA.
There was an intention to deprecate DSA_SIG as well, but we don't have a clear replacement, so that will have to happen separately.

-----

Amends #13187

### CODE REVIEW

**Comment by levitte** [comment]
There's one test error that's du to a lack of MSBLOB / PVK encoders...  this should have been discovered earlier, but a fault in `test/recipes/15-test_rsa.t` had it fail to fail properly.  You gotta appreciate the irony...

**Review by paulidale — COMMENTED** [review]
Your additions/changes look okay to me.
I won't approve since a fair portion of this is mine.

**Comment by levitte** [comment]
This is now pending on #13645

**Comment by t8m** [comment]
Actually as far as I remember DSA_SIG was not intended to be deprecated in 3.0 when we discussed the additional deprecations on the OTC meetings.

**Comment by levitte** [comment]
> Actually as far as I remember DSA_SIG was not intended to be deprecated in 3.0 when we discussed the additional deprecations on the OTC meetings.

Yup, that's even written up in #13121.  The effort in #13187 was a bit over-zealous...  that's ok, it was easy to fix.

**Comment by levitte** [comment]
Rebase, because #13645 got merged

**Comment by levitte** [comment]
Apart from the usual sanitizer, CIs are happy.  Good enough for a review?

File: apps/req.c [hunk 1]
```diff
@@ -626,7 +626,7 @@ int req_main(int argc, char **argv)
                        "         Your key size is %ld! Larger key size may behave not as expected.\n",
                        OPENSSL_RSA_MAX_MODULUS_BITS, newkey);
 
-#ifndef OPENSSL_NO_DSA
+#if !defined(OPENSSL_NO_DSA) || !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline apps/req.c:629]: Why is this change needed?

File: CHANGES.md [hunk 2]
```diff
@@ -504,20 +504,24 @@ OpenSSL 3.0
 
  * All of the low level DSA functions have been deprecated including:
 
+   DSA_SIG_new, DSA_SIG_free, DSA_SIG_get0, DSA_SIG_set0, DSA_new, DSA_free,
```
Comment (t8m) [inline CHANGES.md:507]: DSA_SIG related functions are not deprecated.

File: doc/man3/d2i_X509.pod [hunk 3]
```diff
@@ -51,15 +51,6 @@ d2i_DIRECTORYSTRING,
...
 d2i_DIST_POINT,
 d2i_DIST_POINT_NAME,
-d2i_DSAPrivateKey,
-d2i_DSAPrivateKey_bio,
-d2i_DSAPrivateKey_fp,
-d2i_DSAPublicKey,
-d2i_DSA_PUBKEY,
-d2i_DSA_PUBKEY_bio,
-d2i_DSA_PUBKEY_fp,
-d2i_DSA_SIG,
```
Comment (t8m) [inline doc/man3/d2i_X509.pod:61]: This is not/should not be deprecated.

File: fuzz/asn1.c [hunk 4]
```diff
@@ -335,7 +335,7 @@ int FuzzerTestOneInput(const uint8_t *buf, size_t len)
     DO_TEST_NO_PRINT(DH, d2i_DHparams, i2d_DHparams);
     DO_TEST_NO_PRINT(DH, d2i_DHxparams, i2d_DHxparams);
 #endif
-#ifndef OPENSSL_NO_DSA
+#if !defined(OPENSSL_NO_DSA) || !defined(OPENSSL_NO_DEPRECATED_3_0)
     DO_TEST_NO_PRINT(DSA_SIG, d2i_DSA_SIG, i2d_DSA_SIG);
```
Comment (t8m) [inline fuzz/asn1.c:339]: Should be outside of the deprecated ifdef.

File: doc/man3/d2i_RSAPrivateKey.pod [hunk 5]
```diff
@@ -8,6 +8,15 @@ Any deprecated keypair/params d2i or i2d functions are collected on this page.
...
 =head1 NAME
 
+d2i_DSAPrivateKey,
+d2i_DSAPrivateKey_bio,
+d2i_DSAPrivateKey_fp,
+d2i_DSAPublicKey,
+d2i_DSA_PUBKEY,
+d2i_DSA_PUBKEY_bio,
+d2i_DSA_PUBKEY_fp,
+d2i_DSA_SIG,
```
Comment (t8m) [inline doc/man3/d2i_RSAPrivateKey.pod:18]: Not deprecated.

File: doc/man3/DSA_SIG_new.pod [hunk 6]
```diff
@@ -9,13 +9,21 @@ DSA_SIG_new, DSA_SIG_free - allocate and free DSA signature objects
 
  #include <openssl/dsa.h>
 
+Deprecated since OpenSSL 3.0, can be hidden entirely by defining
```
Comment (t8m) [inline doc/man3/DSA_SIG_new.pod:12]: Those functions are not deprecated.

**Review by t8m — CHANGES_REQUESTED** [review]
There are some leftovers from the attempt to deprecate DSA_SIG.

File: apps/req.c (hunk 1 above)
Comment (levitte) [reply apps/req.c:629]: At this point, that's a very good question

File: apps/dsa.c [hunk 7]
```diff
@@ -171,14 +181,15 @@ int dsa_main(int argc, char **argv)
...
-        dsa = EVP_PKEY_get1_DSA(pkey);
-
-    if (dsa == NULL) {
+    if (pkey == NULL) {
         BIO_printf(bio_err, "unable to load Key\n");
         ERR_print_errors(bio_err);
         goto end;
     }
+    if (!EVP_PKEY_is_a(pkey, "DSA")) {
+        BIO_printf(bio_err, "Not an DSA key\n");
```
[the remark this answers is not in the document]
Comment (levitte) [reply apps/dsa.c:190]: Fixed

File: CHANGES.md (hunk 2 above)
Comment (levitte) [reply CHANGES.md:507]: Fixed

File: doc/man3/d2i_X509.pod (hunk 3 above)
Comment (levitte) [reply doc/man3/d2i_X509.pod:61]: Fixed

File: fuzz/asn1.c (hunk 4 above)
Comment (levitte) [reply fuzz/asn1.c:339]: Fixed

File: doc/man3/d2i_RSAPrivateKey.pod (hunk 5 above)
Comment (levitte) [reply doc/man3/d2i_RSAPrivateKey.pod:18]: Fixed

File: doc/man3/DSA_SIG_new.pod (hunk 6 above)
Comment (levitte) [reply doc/man3/DSA_SIG_new.pod:12]: Fixed

File: apps/req.c (hunk 1 above)
Comment (levitte) [reply apps/req.c]: Undone

**Review by t8m — APPROVED** [review]
LGTM

**Comment by levitte** [comment]
I rebased this to be on top of the newly merged #13661

**Comment by t8m** [comment]
Still approved, just for the record.

**Comment by levitte** [comment]
Merged

d6fff343c85b1fd1c144690b881bf89aa6d049e6 dsa: documentation deprecation changes
575b36ecefca4eff181210ff1eeb3a3dcfbf5456 dsa: fuzzer deprecation changes
47422549da431cf9546a148d916d162e196fcd44 dsa: apps deprecation changes
52c8535a73c575e5ac5f742c35b4fa65ed5df0dc dsa: provider and library deprecation changes
ea7808143d6880db0cb9eb9cf6694c23958d32b0 dsa: add additional deprecated functions to CHANGES entry.
